### PR TITLE
New Addon Bot detection

### PR DESCRIPTION
**Pull request by nupplaphil**
New addon for bot detection.
Using https://github.com/JayBizzle/Crawler-Detect

Solves https://github.com/friendica/friendica/issues/6948

- [x] adding composer.json
- [x] rename to "blockbot"

### CODE REVIEW

**Comment by annando** [comment]
:grin: We are working on the same issue: https://github.com/annando/friendica-addons/blob/blockbot/blockbots/blockbots.php

**Comment by nupplaphil** [comment]
ah :D .. OK I'm closing it :D 

I thought using the suggested library is quite easy. Did you find a cleaner way? Or do you want to avoid a new library?

**Comment by annando** [comment]
I wouldn't hinder you on continue working on your addon. If you want to, please continue. Only thing is that you please return error 403, not 404. I'm just testing the bot detection, that's why the branch is so complicated.

**Comment by nupplaphil** [comment]
Challenge accepted :D

**Comment by nupplaphil** [comment]
@MrPetovan I did it like you said with the composer.

The advancedcontentfilter includes the `vendor` directory, so I added it too. Was the include intended? Just to be sure ;-) (there are now 3.000 lines of code now included because of the vendor directory ;-) )

File: blockbot/composer.json [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+	],
+	"require": {
+		"php": ">=5.6.0",
+		"jaybizzle/crawler-detect": "1.*"
+	},
+	"license": "3-clause BSD license",
+	"minimum-stability": "stable",
+	"config": {
+		"optimize-autoloader": true,
+		"autoloader-suffix": "AdvancedContentFilterAddon",
```
Comment (MrPetovan) [inline blockbot/composer.json]: Please change the autoloader suffix to match your addon.

File: blockbot/blockbot.php [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+	Hook::register('init_1', 'addon/blockbot/blockbot.php', 'blockbot_init_1');
+}
+
+
+function blockbot_uninstall() {
+	Hook::unregister('init_1', 'addon/blockbot/blockbot.php', 'blockbot_init_1');
+}
+
+function blockbot_init_1(App $a) {
+	$crawlerDetect = new CrawlerDetect();
```
Comment (MrPetovan) [inline blockbot/blockbot.php]: Have you tested this code? I would expect an `include __DIR__ . '/vendor/autoload.php';` at some point.

**Comment by annando** [comment]
I don't really like to include a static vendor folder. Thing is: This is a "living" library that is extended constantly. To reliably detect new bots, the library has to be updated all the time.

**Comment by MrPetovan** [comment]
Whether it is in the addon composer or the core composer, both would require a `composer update` run to update the bot list. You would be updating `friendica/friendica-addons` instead of `friendica/friendica`.

**Comment by annando** [comment]
Is there any way that administrators (not the developers!) could update this library on their own, without provoking a GIT problem because of locally changed files?

File: blockbot/blockbot.php (hunk 2 above)
Comment (nupplaphil) [reply blockbot/blockbot.php]: It worked at my local node.. But I'll try it again

**Comment by nupplaphil** [comment]
I added the autoloader require, although it really worked on my local node without it. I faked the "Google Bot" user agent and got a WSOD with 403 :-)

**Comment by nupplaphil** [comment]
> Is there any way that administrators (not the developers!) could update this library on their own, without provoking a GIT problem because of locally changed files?

They should not alter the added sources of the `addon` folder. But afaik there's no possibility to update the patterns without a composer update

**Comment by MrPetovan** [comment]
> I added the autoloader require, although it really worked on my local node without it. I faked the "Google Bot" user agent and got a WSOD with 403 :-)

Did you have the library in your core vendor folder still? That could explain why it's working without the addon autoloader..

**Comment by MrPetovan** [comment]
> Is there any way that administrators (not the developers!) could update this library on their own, without provoking a GIT problem because of locally changed files?

Not if the library doesn't provide a download feature from a remote server. It would be up to us to update the definition file on a regular basis, ideally before each release.

**Comment by nupplaphil** [comment]
> > I added the autoloader require, although it really worked on my local node without it. I faked the "Google Bot" user agent and got a WSOD with 403 :-)
> 
> Did you have the library in your core vendor folder still? That could explain why it's working without the addon autoloader..

Yep, you're right ... Thanks for the hint!

File: blockbot/blockbot.php [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+use Friendica\App;
+use Friendica\Core\Hook;
+use Friendica\Core\System;
+use Jaybizzle\CrawlerDetect\CrawlerDetect;
+
+require_once __DIR__ . DIRECTORY_SEPARATOR . 'vendor' . DIRECTORY_SEPARATOR . 'autoload.php';
+
+function blockbot_install() {
+	Hook::register('init_1', 'addon/blockbot/blockbot.php', 'blockbot_init_1');
```
Comment (MrPetovan) [inline blockbot/blockbot.php]: I suggest using `__FILE__` instead of `'addon/blockbot/blockbot.php'`.

File: blockbot/blockbot.php [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+}
+
+
+function blockbot_uninstall() {
+	Hook::unregister('init_1', 'addon/blockbot/blockbot.php', 'blockbot_init_1');
```
Comment (MrPetovan) [inline blockbot/blockbot.php]: I suggest using `__FILE__` instead of `'addon/blockbot/blockbot.php'`.

File: blockbot/blockbot.php (hunk 3 above)
Comment (nupplaphil) [reply blockbot/blockbot.php]: But `__FILE__` creates a absolute path, would this be a problem? Because currently we're just saving relative paths

Comment (MrPetovan) [reply blockbot/blockbot.php]: It used to be, but we now strip Friendica basedir from the file path if it's absolute.

**Comment by AlfredSK** [comment]
I'm seeing a lot of **403**'s  for `"python-opengraph-jaywink/0.2.0 (+https://github.com/jaywink/python-opengraph)"` and `"Social-Relay/1.6.0-dev - https://github.com/jaywink/social-relay"` now. It this intended, completely unrelated or does the add-on break the Diaspora relay?

**Comment by AlfredSK** [comment]
When I disable the add-on the 403's for these two user agents are gone. So, it seems related to the add-on.

**Comment by AlfredSK** [comment]
And, apparently, it breaks my server/node monitoring. The monitor crawls the nodeinfo every couple of minutes. When it gets a 40x or 50x I'm notified via email about it.

```
--SERVICE-ALERT-------------------
-
- Hostaddress: 134.119.20.10
- Hostname: librabox (J125736)
- Service: Friendica Nodeinfo (M9346)
- - - - - - - - - - - - - - - - -
- State: CRITICAL
- Date: 2019-04-22 15:06:03
- Output: HTTP CRITICAL - Invalid HTTP response received from host on port 443: HTTP/1.1 403 Error 403 - Forbidden
-
----------------------------------
```

**Comment by nupplaphil** [comment]
Seems like the crawler is a little bit more restrictive than necessary...

**Comment by MrPetovan** [comment]
Both are bots, so it makes sense. For me the problem isn't in the library, but in its underlying motivation. You simply can't reliably guess the good/bad intents from any single request, and any attempt to close part or all the incoming traffic based on broad rules will inevitably fail with false positives or false negatives.

**Comment by AlfredSK** [comment]
Yes, I understand that. But there should be a whitelist for the admin then or at least a fat warning that the add-on may break some communication. When the add-on breaks communication with the relay or some other useful remote services it is kind of dangerous to activate it. 
I guess not every admin is looking very closely on the logs. I see a lot of bug reports/support questions coming. :-)

**Comment by MrPetovan** [comment]
A whitelist sounds like a nice addition.

**Comment by nupplaphil** [comment]
Whitelist should be easy
- adding a textbox to the add-on admin page
- save to config `blockbot.whitelist`
- `$crawlerDetector->isCrawler() && !inWhitelist()` 

 :-)

**Comment by tobiasd** [comment]
> New Addon Bot detection

Could you add a descriptive README.md file to the addon explaining what it does and what problems it might raise?

**Comment by annando** [comment]
I have got a whitelist in my corresponding PR. But I think we should create issues at that repository to clean up this list.

**Comment by nupplaphil** [comment]
I'd like to do both (README, Whitelist) in the next 2 days - if there's no other distraction in business ;-) .

**Pull request merged**